src/xdg-autostart-generator/xdg-autostart-service.c

```diff
@@ -0,0 +1,655 @@
+                const char *filename,
+                unsigned line,
+                const char *section,
+                unsigned section_line,
+                const char *lvalue,
+                int ltype,
+                const char *rvalue,
+                void *data,
+                void *userdata) {
+
+        char **s = data;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+        assert(data);
+
+        /* XDG does not allow duplicate definitions. */
+        if (*s) {
+                log_syntax(unit, LOG_ERR, filename, line, 0, "Key %s was defined multiple times, ignoring.", lvalue);
+                return 0;
+        }
+
+        *s = strdup(skip_leading_chars(rvalue, NULL));
+        if (*s == NULL)
+                return log_oom();
+
+        return xdg_unescape_string(unit, filename, line, *s);
+}
+
+static int xdg_config_parse_strv(
+                const char *unit,
+                const char *filename,
+                unsigned line,
+                const char *section,
+                unsigned section_line,
+                const char *lvalue,
+                int ltype,
+                const char *rvalue,
+                void *data,
+                void *userdata) {
+
+        char ***sv = data;
+        char *start;
+        char *end;
+        int r;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+        assert(data);
+
+        /* XDG does not allow duplicate definitions. */
+        if (*sv) {
+                log_syntax(unit, LOG_ERR, filename, line, 0, "Key %s was defined multiple times, ignoring.", lvalue);
+                return 0;
+        }
+
+        *sv = strv_new(NULL);
+        if (*sv == NULL)
+                return log_oom();
+
+        /* We cannot use strv_split because it does not handle escaping correctly. */
+        start = skip_leading_chars(rvalue, NULL);
+
+        for (end = start; *end; end++) {
+                if (*end == '\\') {
+                        /* Move forward, and ensure it is a valid escape. */
+                        end++;
+                        if (strchr("sntr\\;", *end) == NULL) {
+                                log_syntax(unit, LOG_ERR, filename, line, 0, "Undefined escape sequence \\%c.", *end);
+                                return 0;
+                        }
+                        continue;
+                }
+
+                if (*end == ';') {
+                        _cleanup_free_ char *copy = NULL;
+
+                        copy = strndup(start, end - start);
+                        if (copy == NULL)
+                                return log_oom();
+                        r = xdg_unescape_string(unit, filename, line, copy);
+                        if (r < 0)
+                                return r;
+                        r = strv_consume(sv, copy);
+                        if (r < 0)
+                                return r;
+                        copy = NULL;
```

Comment:
strv_consume is a bit weird, it's a wrapper around strv_push() for the sole reason to destroy the copied string on failure.  but in your code you leave copy set on failure, thus a bad memory access is triggered. Consider using:

```
r = strv_consume(sv, TAKE_PTR(copy));
```

which makes sure `copy` is invalidated before the `if`